wxConnection::OnDisconnect() should schedule the wxConnection object for deletion instead of deleting it immediately



Signed-off-by: Anders Larsen <al@alarsen.net>
---

 include/wx/ipcbase.h   |    2 +-
 src/common/ipcbase.cpp |   20 ++++++++++++++++++++
 2 files changed, 21 insertions(+), 1 deletions(-)


diff --git a/include/wx/ipcbase.h b/include/wx/ipcbase.h
index 3918547..5c39c0b 100644
--- a/include/wx/ipcbase.h
+++ b/include/wx/ipcbase.h
@@ -177,7 +177,7 @@ public:
       { return false; }
 
   // Callbacks to BOTH
-  virtual bool OnDisconnect() { delete this; return true; }
+  virtual bool OnDisconnect();
 
 
   // return true if this is one of the formats used for textual data
diff --git a/src/common/ipcbase.cpp b/src/common/ipcbase.cpp
index 64c6c1f..2f078a9 100644
--- a/src/common/ipcbase.cpp
+++ b/src/common/ipcbase.cpp
@@ -19,6 +19,7 @@
 #ifndef WX_PRECOMP
 #endif
 
+#include "wx/apptrait.h"
 #include "wx/ipcbase.h"
 
 IMPLEMENT_CLASS(wxServerBase, wxObject)
@@ -65,6 +66,25 @@ wxConnectionBase::~wxConnectionBase()
     delete [] m_buffer;
 }
 
+
+bool wxConnectionBase::OnDisconnect()
+{
+    // schedule this object for deletion
+    wxAppTraits *traits = wxTheApp ? wxTheApp->GetTraits() : NULL;
+    if ( traits )
+    {
+        // let the traits object decide what to do with us
+        traits->ScheduleForDestroy(this);
+    }
+    else // no app or no traits
+    {
+        // in wxBase we might have no app object at all, don't leak memory
+        delete this;
+    }
+    return true;
+}
+
+
 /* static */
 wxString wxConnectionBase::GetTextFromData(const void* data,
                                            size_t size,
